File: exercise/ts/day02/src/fizzbuzz.ts
import {none, Option, some} from "fp-ts/Option";

type Config = {
    mapping: Map<number, string>;
    min: number;
    max: number;
}
export const fizzbuzz = ({mapping, min, max}: Config) => (input: number): Option<string> =>
    isOutOfRange({min, max})(input)
        ? none
        : some(convertSafely(mapping)(input));

const convertSafely = (mapping: Map<number, string>) => (input: number): string => {
    const values: string[] = [];
    for (const [divisor, value] of mapping) {
        if (is(divisor, input)) {
            values.push(value);
        }
    }
    return values.length ? values.join('') : input.toString();
};

const is = (divisor: number, input: number): boolean => input % divisor === 0;
const isOutOfRange = ({min, max}: {min:number; max: number;}) => (input: number): boolean => input < min || input > max;

File: solution/day02/src/main/kotlin/games/FizzBuzz.kt
package games

import arrow.core.None
import arrow.core.Option
import arrow.core.Some

const val MIN = 1
const val MAX = 100
typealias MatchingDivisors = Collection<String>
typealias Map = kotlin.collections.Map<Int, String>

object FizzBuzz {
    fun convert(mapping: Map, input: Int): Option<String> = when {
        isOutOfRange(input) -> None
        mapping.isEmpty() -> None
        else -> Some(convertSafely(mapping, input))
    }

    private fun isOutOfRange(input: Int) = input < MIN || input > MAX

    private fun convertSafely(mapping: Map, input: Int): String =
        mapping.matchingDivisors(input)
            .values
            .toResult(input)

    private fun Map.matchingDivisors(input: Int) = filter { (key, _) -> `is`(key, input) }
    private fun `is`(divisor: Int, input: Int): Boolean = input % divisor == 0

    private fun MatchingDivisors.toResult(input: Int) = when {
        any() -> joinToString("")
        else -> input.toString()
    }
}